File: documentation/manual/working/commonGuide/akka/code/scalaguide/akka/typed/oo/AppModule.scala
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package scalaguide.akka.typed.oo

// #oo-app-module
import akka.actor.ActorSystem
import akka.actor.typed.scaladsl.adapter._
import akka.actor.typed.ActorRef
import com.google.inject.AbstractModule
import com.google.inject.Provider
import com.google.inject.TypeLiteral
import javax.inject.Inject
import play.api.Configuration
import play.api.libs.concurrent.AkkaGuiceSupport

object AppModule extends AbstractModule with AkkaGuiceSupport {
  override def configure() = {
    bindTypedActor(HelloActor.create(), "hello-actor")
    bind(new TypeLiteral[ActorRef[ConfiguredActor.GetConfig]]() {})
      .toProvider(classOf[ConfiguredActorProvider])
      .asEagerSingleton()
  }

  private class ConfiguredActorProvider @Inject() (
      actorSystem: ActorSystem,
      configuration: Configuration,
  ) extends Provider[ActorRef[ConfiguredActor.GetConfig]] {
    def get() = {
      val behavior = ConfiguredActor.create(configuration)
      actorSystem.spawn(behavior, "configured-actor")
    }
  }
}
// #oo-app-module

File: documentation/manual/working/commonGuide/akka/code/javaguide/akka/typed/AppModule.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package javaguide.akka.typed;

// #oo-app-module
import akka.actor.ActorSystem;
import akka.actor.typed.javadsl.Adapter;
import akka.actor.typed.ActorRef;
import com.google.inject.AbstractModule;
import com.google.inject.Provider;
import com.google.inject.TypeLiteral;
import com.typesafe.config.Config;
import javax.inject.Inject;
import play.api.Configuration;
import play.api.libs.concurrent.AkkaGuiceSupport;

public class AppModule extends AbstractModule {

  @Override
  protected void configure() {
    bind(new TypeLiteral<ActorRef<HelloActor.SayHello>>() {})
        .toProvider(
            new Provider<ActorRef<HelloActor.SayHello>>() {
              @Inject ActorSystem actorSystem;

              @Override
              public ActorRef<HelloActor.SayHello> get() {
                return Adapter.spawn(actorSystem, HelloActor.create(), "hello-actor");
              }
            })
        .asEagerSingleton();
    bind(new TypeLiteral<ActorRef<ConfiguredActor.GetConfig>>() {})
        .toProvider(
            new Provider<ActorRef<ConfiguredActor.GetConfig>>() {
              @Inject ActorSystem actorSystem;
              @Inject Config config;

              @Override
              public ActorRef<ConfiguredActor.GetConfig> get() {
                return Adapter.spawn(
                    actorSystem, ConfiguredActor.create(config), "configured-actor");
              }
            })
        .asEagerSingleton();
  }
}
// #oo-app-module
